variant_remapping_tools/reverse_strand.py

```diff
@@ -72,43 +20,116 @@
+                and (start + operator <= (read_var_pos + local_region_size))
+                and (start + operator != read_var_pos)):
             if cig_list[operator] == 0:  # Match or mismatch
                 perf_counter += 1  # Increase the counter for perfect local region
-    if (AS <= int(score_cutoff)):
-        primary_poor_count += 1
-    elif (AS - XS < diff_cutoff):
-        gap_small_count += 1
-    elif (perf_counter != 2*local_region_size):
-        context_bad_count += 1
+    return varpos, perf_counter, local_region_size
+
+
+def is_read_valid(read, counter, flank_length, score_cutoff, diff_cutoff):
+    if read.is_secondary:  # We only want to deal with primary reads
+        return False
+    counter['total'] += 1
+    if read.is_unmapped:  # Can be decoded with bitwise flag with & 4 (4 means unmapped)
+        counter['unmapped'] += 1
+        return False
+    # Getting AS and XS:
+    AS = read.get_tag("AS")
+    try:
+        XS = read.get_tag("XS")  # Some variants don't have secondary alignments, which throws an error
+    except KeyError:
+        XS = -flank_length  # Set an arbitrary low value for the "artificial" secondary alignment
+    if AS <= int(score_cutoff):
+        counter['primary_poor'] += 1
+        return False
+    if AS - XS < diff_cutoff:
+        counter['gap_small'] += 1
+        return False
+    varpos, perf_counter, local_region_size = calculate_variant_position(read, flank_length)
+    if perf_counter != 2 * local_region_size:
+        counter['context_bad'] += 1
+        return False
     # Filter out AS's that are too low and XS's that are too close to AS, and those with a non-perfect local alignment:
-    if (AS > int(score_cutoff)) and (AS - XS >= diff_cutoff) and (perf_counter == 2*local_region_size):
-        remapped_count += 1
-        # Write it all to the file:
-        outfile.write("%s\t%s\t%s\t%s\t%s\t%s\t%s\n" % (read.reference_name,  varpos, info[4], nucl, info[5], info[6],\
-                                                        info[7]))
-        # Store old reference allele:
-        old_ref_alleles.write("%s\n" % info[2])
-outfile.close()
-old_ref_alleles.close()
-print(total_count)
-print(unmapped_count)
-print(primary_poor_count)
-print(gap_small_count)
-print(context_bad_count)
-print(remapped_count)
-print("% of variants rejected for:")
-print("Unmapped: {:.2%}".format(unmapped_count/total_count))
-print("Primary alignment too poor: {:.2%}".format(primary_poor_count/total_count))
-print("Primary and secondary alignments too close: {:.2%}".format(gap_small_count/total_count))
-print("Local region around variant too poorly aligned: {:.2%}".format(context_bad_count/total_count))
\ No newline at end of file
+    counter['remapped'] += 1
+    return True
+
+
+def process_bam_file(bam_file_path, output_file, old_ref_allele_output_file, flank_length, score_perc, diff_AS_XS_perc):
+    bamfile = pysam.AlignmentFile(bam_file_path, 'rb')
+    outfile = open(output_file, 'w')
```

Comment:
Suggest using the [with expression](https://docs.python.org/2.5/whatsnew/pep-343.html) to close write file handles in case of errors.

Reply:
Didn't want to change anything but I guess that won't change the logic.
Done